cluster/ci/origin/pipelines/test-build/Jenkinsfile

```diff
@@ -0,0 +1,127 @@
+    string(name: "PULL_NUMBER", defaultValue:"${env.PULL_NUMBER ?: ''}")
+    string(name: "PULL_PULL_SHA", defaultValue:"${env.PULL_PULL_SHA ?: ''}")
+    string(name: "RELEASE_SRC_URL", defaultValue: "${env.RELEASE_SRC_URL ?: 'https://github.com/openshift/release.git'}")
+    string(name: "RELEASE_SRC_REF", defaultValue: "${env.RELEASE_SRC_REF ?: 'master'}")
+    string(name: "RELEASE_LIBRARY_URL", defaultValue: "${library_url}")
+    string(name: "RELEASE_LIBRARY_REF", defaultValue: "${library_ref}")
+  }
+
+  stages {
+    stage ("Configure Variables") {
+      steps {
+        script {
+          this._buildName = buildName(this)
+          this._pushSecret = dockerCfgSecret(this, "builder")
+          this._jobId = "test-build-${this._buildName}-${env.BUILD_NUMBER}"
+        }
+        echo "Build name is ${this._buildName}"
+      }
+    }
+    stage ("Create ImageStream") {
+      when {
+        not { expression { return imageStreamExists(this, "${this._buildName}") } }
+      }
+      steps {
+        echo "Creating image stream ${this._buildName}"
+        applyTemplate(this, "${templateBase}/output-imagestream.yaml", 
+          "BUILD_NAME=${this._buildName}")
+      }
+    }
+    stage("Clone") {
+      when {
+        not { expression { return imageStreamTagExists(this, "${this._buildName}", "src") } }
+      }
+      steps {
+        script {
+          jitter(this, 0,30)
+          def activeBuilds = filterObjects(this, "builds", ["build":"${this._buildName}"], { obj -> 
+            if (obj.metadata.labels != null && obj.metadata.labels["template"] == "clone-source") {
+              if (obj.status.phase == "New" || obj.status.phase == "Pending" || obj.status.phase == "Running") {
+                return true
+              }
+            }
+            return false
+          })
+          if (activeBuilds.size() == 0) {
+            runAndWaitForBuild(this, "", {
+              return createTemplate(this, "${templateBase}/clone-source.yaml",
+                "RELEASE_SRC_REF=${params.RELEASE_SRC_REF}",
+                "RELEASE_SRC_URL=${params.RELEASE_SRC_URL}", 
+                "PUSH_SECRET_NAME=${this._pushSecret}",
+                "BUILD_NAME=${this._buildName}",
+                "PULL_REFS=${params.PULL_REFS}",
+                "PULL_NUMBER=${params.PULL_NUMBER}",
+                "BUILD_NUMBER=${env.BUILD_NUMBER}",
+                "REPO_OWNER=${params.REPO_OWNER}",
+                "REPO_NAME=${params.REPO_NAME}",
+                "JOB_ID=${this._jobId}")
+            })
+          } else {
+            echo "Skipping source build for ${this._buildName}. Another build is already running: ${activeBuilds[0].metadata.name}"
+          }
+        }
+      }
+    }
+    stage("Compile") {
+      when {
+        not { expression { return imageStreamTagExists(this, "${this._buildName}", "test-binaries") } }
+      }
+      steps {
+        script {
+          jitter(this, 0,30)
```

Comment:
why does the main build pipe not use this pattern?

Reply:
because it really doesn't get us much and I want us to write a better utility to launch monitor builds from the corresponding test.